mlir/lib/IR/AttributeDetail.h

```diff
@@ -411,12 +414,56 @@ class DistinctAttributeAllocator {
   /// Allocates a distinct attribute storage using a thread local bump pointer
   /// allocator to enable synchronization free parallel allocations.
   DistinctAttrStorage *allocate(Attribute referencedAttr) {
-    return new (allocatorCache.get().Allocate<DistinctAttrStorage>())
-        DistinctAttrStorage(referencedAttr);
+    if (useSynchronizedAllocator && !useThreadLocalAllocator) {
+      std::scoped_lock<std::mutex> lock(allocatorMutex);
+      return allocateImpl(referencedAttr);
+    }
+    if (!useSynchronizedAllocator)
+      return allocateImpl(referencedAttr);
+    llvm_unreachable(
+        "Cannot use both a synchronised and thread_local allocator!");
+  }
+
+  /// Sets flags to use thread local bump pointer allocators or a single
+  /// non-thread safe bump pointer allocator depending on if multi-threading is
+  /// enabled. Use this to disable the use of thread local storage and bypass
+  /// thread safety synchronization overhead.
+  void disableMultiThreading(bool disable = true) {
+    disableThreadLocalStorage(disable);
+    useSynchronizedAllocator = !disable;
```

Comment:
I would avoid modifying both flags in the disableMultiThreading / disableThreadLocalStorage functions. Is this needed to avoid the unreachable? Can't we just modify the flags separately and then have the necessary logic in the allocate function?